### PR TITLE
Add AVX2 version of adler32

### PR DESCRIPTION
**Pull request by brianpopow**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

This PR adds a AVX2 version of adler method.

#### Benchmark results:
```
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19043.1526 (21H1/May2021Update)
Intel Core i7-6700K CPU 4.00GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET SDK=6.0.200
  [Host]     : .NET 5.0.14 (5.0.1422.5710), X64 RyuJIT
  DefaultJob : .NET 5.0.14 (5.0.1422.5710), X64 RyuJIT


|      Method |       Mean |   Error |  StdDev |
|------------ |-----------:|--------:|--------:|
|    AdlerSse |   160.4 ns | 1.10 ns | 0.92 ns |
|    AdlerAvx |   109.6 ns | 0.52 ns | 0.49 ns |
| AdlerScalar | 1,557.8 ns | 4.71 ns | 4.40 ns |
```

### CODE REVIEW

**Comment by JimBobSquarePants** [comment]
I kept meaning to migrate this across.

https://github.com/SixLabors/ZlibStream/blob/main/src/ZlibStream/Adler32.cs

Compression  is currently broken in that repo (to a point I’m going to have to start again from the beginning of my fork and re-optimize with better tests) but Decompression works well and benchmarked faster at times than the runtime implementation. There’s some fast code in there.

**Comment by brianpopow** [comment]
> I kept meaning to migrate this across.
> 
> https://github.com/SixLabors/ZlibStream/blob/main/src/ZlibStream/Adler32.cs
> 
> Compression is currently broken in that repo (to a point I’m going to have to start again from the beginning of my fork and re-optimize with better tests) but Decompression works well and benchmarked faster at times than the runtime implementation. There’s some fast code in there.

Ok, I was not aware of this. Will close this PR then.

**Comment by JimBobSquarePants** [comment]
I’d still compare performance though. Your Adler might still be faster!

**Comment by brianpopow** [comment]
> I’d still compare performance though. Your Adler might still be faster!

The implementation from [SixlaborsZlib](https://github.com/SixLabors/ZlibStream/blob/main/src/ZlibStream/Adler32.cs) is faster:

```
|      Method |        Mean |    Error |   StdDev |
|------------ |------------:|---------:|---------:|
|   AdlerSse2 |   167.95 ns | 1.974 ns | 1.750 ns |
|  AdlerSsse3 |    87.87 ns | 1.716 ns | 3.051 ns |
|    AdlerAvx |   111.50 ns | 0.761 ns | 0.712 ns |
| AdlerScalar | 1,566.36 ns | 5.335 ns | 4.730 ns |
```

After sleeping one night over it, I thought it might be still worth considering merging the AVX version of this PR to ImageSharp until we can make use of https://github.com/SixLabors/ZlibStream/blob/main/src/ZlibStream/Adler32.cs. I think it will take some time until we can make [SixLabors/ZlibStream](https://github.com/SixLabors/ZlibStream) production ready. @JimBobSquarePants what do you think?

**Comment by JimBobSquarePants** [comment]
I don’t know when I’ll get time to work exclusively on ZlibStream. It’s a minimum of a few weeks full time work to incrementally implement the optimisations again unless someone can spot the issue. So yeah, merge ahead if it’s easier to do so

**Comment by brianpopow** [comment]
Ok then I think reopening this again makes sense.

**Review by JimBobSquarePants — APPROVED** [review]
LGTM 👍

File: src/ImageSharp/Compression/Zlib/Adler32.cs [hunk 1]
```diff
@@ -83,18 +90,16 @@ private static unsafe uint CalculateSse(uint adler, ReadOnlySpan<byte> buffer)
...
             uint length = (uint)buffer.Length;
-            uint blocks = length / BLOCK_SIZE;
-            length -= blocks * BLOCK_SIZE;
+            uint blocks = length / BlockSize;
+            length -= blocks * BlockSize;
 
             int index = 0;
             fixed (byte* bufferPtr = buffer)
             {
                 fixed (byte* tapPtr = Tap1Tap2)
```
Comment (gfoidl) [inline src/ImageSharp/Compression/Zlib/Adler32.cs:100]:
```suggestion
            fixed (byte* bufferPtr = &MemoryMarshal.GetReference(buffer))
            {
                fixed (byte* tapPtr = &MemoryMarshal.GetReference(Tap1Tap2))
```
For `buffer` there's the empty-check done before, so no need to repeat it (implicitely by GetPinnableReference which C# generates for the fixed statement here).
For `Tap1Tag2` it's a "constant" we know for sure it's not-empty.

Note: also add the namespace needed for _MemoryMarshal_.

File: src/ImageSharp/Compression/Zlib/Adler32.cs [hunk 2]
```diff
@@ -164,45 +169,127 @@ private static unsafe uint CalculateSse(uint adler, ReadOnlySpan<byte> buffer)
...
+        // Based on: https://github.com/zlib-ng/zlib-ng/blob/develop/arch/x86/adler32_avx2.c
+        [MethodImpl(InliningOptions.HotPath | InliningOptions.ShortMethod)]
+        public static unsafe uint CalculateAvx2(uint adler, ReadOnlySpan<byte> buffer)
+        {
+            uint s1 = adler & 0xFFFF;
+            uint s2 = (adler >> 16) & 0xFFFF;
+            uint length = (uint)buffer.Length;
 
-                        s2 %= BASE;
+            fixed (byte* bufferPtr = buffer)
```
Comment (gfoidl) [inline src/ImageSharp/Compression/Zlib/Adler32.cs:188]:
```suggestion
            fixed (byte* bufferPtr = &MemoryMarshal.GetReference(buffer))
```
But maybe it's better to have the pinning in the caller method above, and pass the pointer to these methods. Should help the JIT for inlining-decisions, as (especially older runtimes) won't inline methods which pin variables.

**Comment by brianpopow** [comment]
> LGTM 👍

ok lets wait for this to be merged when 2.0.1 is done. I think only hotfixes should go into 2.0.1

**Comment by brianpopow** [comment]
@JimBobSquarePants since we are skipping 2.0.1, this can now also get merged into the `main` branch, right?

**Comment by JimBobSquarePants** [comment]
Yeah, merge away 👍

**Comment by brianpopow** [comment]
> I kept meaning to migrate this across.
> 
> https://github.com/SixLabors/ZlibStream/blob/main/src/ZlibStream/Adler32.cs
> 
> Compression is currently broken in that repo ...

@JimBobSquarePants: The tests in this repo seem to pass and there are encode tests. How did you notice its not working as it should?

**Comment by JimBobSquarePants** [comment]
Let me check my local copy and push failing tests. It’s definitely broken though not related to adler

**Pull request merged**